### PR TITLE
Add events to cso status updates for existing status and added in PR 838

### PR DESCRIPTION
**Pull request by ghanshyam-11**
## Pull request checklist

```release-notes

```

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature Enhancement
- [ ] Test Automation 
- [ ] Code Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Community Operator listing 
- [ ] Other (please describe): 

## What is the current behavior?

-

## What is the new behavior?

-

## How risky is this change?
- [ ] Small, isolated change
- [ ] Medium, requires regression testing
- [ ] Large, requires functional and regression testing

### CODE REVIEW

File: operator/controllers/csiscaleoperator_controller.go [hunk 1]
```diff
@@ -285,13 +265,20 @@ func (r *CSIScaleOperatorReconciler) Reconcile(ctx context.Context, req ctrl.Req
...
+			metav1.ConditionFalse, string(csiv1.CSINotConfigured), err.Error(),
+		)
 		return reconcile.Result{}, err
 	}
 
 	if !cmExists || clustersStanzaModified {
 		err = ValidateCRParams(instance)
 		if err != nil {
-			logger.Error(fmt.Errorf("CR validation for driver configuration params failed"), "")
+			message := "CR validation for driver configuration params failed."
```
Comment (amdabhad) [inline operator/controllers/csiscaleoperator_controller.go:277]: we don't need to set status here as function ValidateCRParams already does that.

File: operator/controllers/csiscaleoperator_controller.go [hunk 2]
```diff
@@ -302,6 +289,9 @@ func (r *CSIScaleOperatorReconciler) Reconcile(ctx context.Context, req ctrl.Req
 		if err != nil {
 			message := "Error in getting connectors"
 			logger.Error(err, message)
+			SetStatusConditionAndEvent(instance, r.Recorder, corev1.EventTypeWarning, string(config.StatusConditionSuccess),
```
Comment (amdabhad) [inline operator/controllers/csiscaleoperator_controller.go:292]: same here, getSpectrumScaleConnectors already sets the status

File: operator/controllers/csiscaleoperator_controller.go [hunk 3]
```diff
@@ -315,6 +305,10 @@ func (r *CSIScaleOperatorReconciler) Reconcile(ctx context.Context, req ctrl.Req
 		//and returns the absolute path of symlink directory in the primary fileset.
 		symlinkDirPath, err = r.handlePrimaryFSandFileset(instance)
 		if err != nil {
+			message := "Error checking primary FS and primary fileset"
```
Comment (amdabhad) [inline operator/controllers/csiscaleoperator_controller.go:308]: similarly, remove status setting from here

File: operator/controllers/csiscaleoperator_controller.go [hunk 4]
```diff
@@ -341,6 +335,9 @@ func (r *CSIScaleOperatorReconciler) Reconcile(ctx context.Context, req ctrl.Req
 		logger.Info("Some of the cluster fields of CSIScaleOperator instance are changed, so restarting node plugin pods")
 		err = r.handleDriverRestart(instance)
 		if err != nil {
+			SetStatusConditionAndEvent(instance, r.Recorder, corev1.EventTypeWarning, string(config.StatusConditionSuccess),
```
Comment (amdabhad) [inline operator/controllers/csiscaleoperator_controller.go:338]: similarly, remove status setting from here

File: operator/controllers/csiscaleoperator_controller.go [hunk 5]
```diff
@@ -201,13 +201,9 @@ func (r *CSIScaleOperatorReconciler) Reconcile(ctx context.Context, req ctrl.Req
...
 		message := "Couldn't add Finalizer"
 		logger.Error(err, message)
-		// TODO: Add event.
-		meta.SetStatusCondition(&crStatus.Conditions, metav1.Condition{
-			Type:    string(config.StatusConditionSuccess),
-			Status:  metav1.ConditionFalse,
-			Reason:  string(csiv1.ResourceUpdateError),
-			Message: message,
-		})
+		SetStatusConditionAndEvent(instance, r.Recorder, corev1.EventTypeWarning, string(config.StatusConditionSuccess),
```
Comment (amdabhad) [inline operator/controllers/csiscaleoperator_controller.go:204]: let's rename the function name `SetStatusConditionAndEvent` something like `SetStatusAndRaiseEvent` ?

File: operator/controllers/csiscaleoperator_controller.go [hunk 6]
```diff
@@ -285,13 +265,20 @@ func (r *CSIScaleOperatorReconciler) Reconcile(ctx context.Context, req ctrl.Req
 
 	cmExists, clustersStanzaModified, err := r.isClusterStanzaModified(req.Namespace, instance)
 	if err != nil {
+		SetStatusConditionAndEvent(instance, r.Recorder, corev1.EventTypeWarning, string(config.StatusConditionSuccess),
```
Comment (amdabhad) [inline operator/controllers/csiscaleoperator_controller.go:268]: similarly, remove status setting from here

**Comment by amdabhad** [comment]
There are many other places where we are setting cso status with `meta.SetStatusCondition` - it should be taken care by the new function `SetStatusConditionAndEvent` in all places.

File: operator/controllers/csiscaleoperator_controller.go (hunk 5 above)
Comment (ghanshyam-11) [reply operator/controllers/csiscaleoperator_controller.go:204]: addressed!

File: operator/controllers/csiscaleoperator_controller.go (hunk 6 above)
Comment (ghanshyam-11) [reply operator/controllers/csiscaleoperator_controller.go:268]: Addressed!

File: operator/controllers/csiscaleoperator_controller.go (hunk 1 above)
Comment (ghanshyam-11) [reply operator/controllers/csiscaleoperator_controller.go:277]: Addressed!

File: operator/controllers/csiscaleoperator_controller.go (hunk 2 above)
Comment (ghanshyam-11) [reply operator/controllers/csiscaleoperator_controller.go:292]: Addressed!

File: operator/controllers/csiscaleoperator_controller.go (hunk 3 above)
Comment (ghanshyam-11) [reply operator/controllers/csiscaleoperator_controller.go:308]: Addressed!

File: operator/controllers/csiscaleoperator_controller.go [hunk 7]
```diff
@@ -291,7 +271,11 @@ func (r *CSIScaleOperatorReconciler) Reconcile(ctx context.Context, req ctrl.Req
 	if !cmExists || clustersStanzaModified {
 		err = ValidateCRParams(instance)
 		if err != nil {
-			logger.Error(fmt.Errorf("CR validation for driver configuration params failed"), "")
+			message := "CR validation for driver configuration params failed."
+			logger.Error(fmt.Errorf(message+"%w", err), "")
+			SetStatusAndRaiseEvent(instance, r.Recorder, corev1.EventTypeWarning, string(config.StatusConditionSuccess),
```
Comment (amdabhad) [inline operator/controllers/csiscaleoperator_controller.go:276]: we don't need to set status here again!

**Pull request closed without merging**